Framework/API/src/IPeakFunction.cpp

```diff
@@ -169,21 +170,40 @@ void IPeakFunction::setPeakRadius(int r) const {
   }
 }
 
-/// Returns the integral intensity of the peak function, using the peak radius
-/// to determine integration borders.
-double IPeakFunction::intensity() const {
-  auto interval = getDomainInterval();
+void IPeakFunction::setParameter(size_t i, const double &value, bool explicitlySet) {
+  m_parameterContextDirty = true;
+  ParamFunction::setParameter(i, value, explicitlySet);
+}
 
-  PeakFunctionIntegrator integrator;
-  IntegrationResult result = integrator.integrate(*this, interval.first, interval.second);
+void IPeakFunction::setParameter(const std::string &name, const double &value, bool explicitlySet) {
+  m_parameterContextDirty = true;
+  ParamFunction::setParameter(name, value, explicitlySet);
+}
 
-  if (!result.success) {
-    return 0.0;
-  }
+// integrate based on dirty parameters then cache the result
+IntegrationResultCache IPeakFunction::integrate() const {
+  if (!integrationResult || m_parameterContextDirty) {
+    auto const interval = getDomainInterval();
+
+    PeakFunctionIntegrator integrator;
 
-  return result.result;
+    auto const result = integrator.integrate(*this, interval.first, interval.second);
+    if (result.success)
+      integrationResult = boost::make_shared<IntegrationResultCache>(result.result, result.error);
+    else
+      integrationResult = boost::make_shared<IntegrationResultCache>(std::nan(""), std::nan(""));
+    m_parameterContextDirty = false;
+  }
+  return *integrationResult;
 }
 
+/// Returns the integral intensity of the peak function, using the peak radius
+/// to determine integration borders.
+double IPeakFunction::intensity() const { return integrate().first; }
+
+/// Returns the uncertainty associated to the integral intensity of the peak function
+double IPeakFunction::intensityError() const { return integrate().second; }
```

Comment:
I'm afraid this is *not* the uncertainty associated to the intensity because `PeakFunctionIntegrator::integrate()` doesn't use the uncertainty in the optimized fit peak parameters. See section "Description" of [PD234](https://code.ornl.gov/sns-hfir-scse/diffraction/powder/powder-diffraction/-/issues/234#description) for the steps needed to obtain a numerical estimation of the uncertainty in the integrated intensity that takes into account the uncertainty in the optimized fitting parameters.  
In the meantime, I believe it's better to return `nan` than return an incorrect result